packages/eslint-plugin/src/rules/prefer-find.ts

```diff
@@ -0,0 +1,313 @@
+  nullThrows,
+} from '../util';
+
+export default createRule({
+  name: 'prefer-find',
+  meta: {
+    docs: {
+      description:
+        'Enforce the use of Array.prototype.find() over Array.prototype.filter() followed by [0] when looking for a single result',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      preferFind: 'Prefer .find(...) instead of .filter(...)[0].',
+      preferFindSuggestion: 'Use .find(...) instead of .filter(...)[0].',
+    },
+    schema: [],
+    type: 'suggestion',
+    hasSuggestions: true,
+  },
+
+  defaultOptions: [],
+
+  create(context) {
+    const globalScope = getScope(context);
+    const services = getParserServices(context);
+    const checker = services.program.getTypeChecker();
+
+    interface FilterExpressionData {
+      isBracketSyntaxForFilter: boolean;
+      filterNode: TSESTree.Node;
+    }
+
+    function parseIfArrayFilterExpression(
+      expression: TSESTree.Expression,
+    ): FilterExpressionData | undefined {
+      if (expression.type === AST_NODE_TYPES.SequenceExpression) {
+        // Only the last expression in (a, b, [1, 2, 3].filter(condition))[0] matters
+        const lastExpression = nullThrows(
+          expression.expressions.at(-1),
+          'Expected to have more than zero expressions in a sequence expression',
+        );
+        return parseIfArrayFilterExpression(lastExpression);
+      }
+
+      if (expression.type === AST_NODE_TYPES.ChainExpression) {
+        return parseIfArrayFilterExpression(expression.expression);
+      }
+
+      // Check if it looks like <<stuff>>(...), but not <<stuff>>?.(...)
+      if (
+        expression.type === AST_NODE_TYPES.CallExpression &&
+        !expression.optional
+      ) {
+        const callee = expression.callee;
+        // Check if it looks like <<stuff>>.filter(...) or <<stuff>>['filter'](...),
+        // or the optional chaining variants.
+        if (callee.type === AST_NODE_TYPES.MemberExpression) {
+          const isBracketSyntaxForFilter = callee.computed;
+          if (isStaticMemberAccessOfValue(callee, 'filter', globalScope)) {
+            const filterNode = callee.property;
+
+            const filteredObjectType = getConstrainedTypeAtLocation(
+              services,
+              callee.object,
+            );
+
+            // As long as the object is an array or an optional chain on a
+            // nullable array, this is an Array.prototype.filter expression.
+            if (isArrayish(filteredObjectType)) {
+              return {
+                isBracketSyntaxForFilter,
+                filterNode,
+              };
+            }
+          }
+        }
+      }
+
+      return undefined;
+    }
+
+    /**
+     * Tells whether the type is a possibly nullable array/tuple or union thereof.
+     */
+    function isArrayish(type: Type): boolean {
+      let isAtLeastOneArrayishComponent = false;
+      for (const unionPart of tsutils.unionTypeParts(type)) {
```

Comment:
[maybe bug] Perhaps we should check for intersections here too:

This code is not reported (but I think it should be!)
_[playground](https://deploy-preview-8216--typescript-eslint.netlify.app/play/#ts=5.3.3&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3rGAueA3tvgIzwC%2BA2gLrwBkh8ARvgEwU0BQnOAdAGYBLCBhAwAFOICU8ALwA%2BeBhgBXEFMoAGakA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0Wlq%2BYATdFF7QA9tEjgwAXxCygA&tsconfig=&tokens=false)_

```ts
declare const arr: { a: 1 }[] & { b: 2 }[]

arr.filter(() => true)[0]
```

+1 on https://github.com/typescript-eslint/typescript-eslint/pull/8250#discussion_r1471943792

Reply:
Added support for this.